### PR TITLE
Add scenario for non allowed ops 

### PR DESCRIPTION
**Pull request by czepluch**
Fixes: https://github.com/raiden-network/raiden/issues/5210

### CODE REVIEW

**Comment by LefterisJP** [comment]
I will review this but from the description of the scenario it does things which I would expect should be tested by unit and integration tests. Why do we also need a scenario for those?

**Comment by czepluch** [comment]
The main reason is that we'd be doing this during manual testing anyway, but with a scenario doing it by using the API, we do it automated in stead. I was asked to go through the old manual tests that we performed for Red Eyes and then see if there were tests we did there that isn't covered in scenarios yet, hence this one was created.

**Comment by LefterisJP** [comment]
So I reviewed this and really don't see a reason to be testing this functionality in a scenario. My reasoning is:

- It ony tests the API of a **single raiden instance**
- There is no synchronization/collaboration between different nodes. 
- For each one of the assertions of this tests exists a corresponding integration test in the raiden client repo.

I would not want to have redundant scenarios in our scenarios run. They already take quite long to run.

**Comment by czepluch** [comment]
Good points. Let me discuss this with Christian when he is back and then I will close the PR if he agrees.

**Comment by czepluch** [comment]
> For each one of the assertions of this tests exists a corresponding integration test in the raiden client repo.

@LefterisJP can you link the corresponding tests? I just want to make sure all of the cases are covered.

**Comment by LefterisJP** [comment]
@czepluch just check our rest api integration tests: https://github.com/raiden-network/raiden/blob/develop/raiden/tests/integration/api/test_restapi.py

If something is missing we can add it there.

**Comment by Dominik1999** [comment]
@LefterisJP can you verify that every test in that scenario is indeed tested in https://github.com/raiden-network/raiden/blob/develop/raiden/tests/integration/api/test_restapi.py

It is our job to make sure that our unit/integration tests test this behaviour. There should be no need to test this in the scenario player.

**Comment by LefterisJP** [comment]
@Dominik1999 Sure. A good indicator of what is covered can always be seen through latest coverage report: https://codecov.io/gh/raiden-network/raiden/src/develop/raiden/api/rest.py

Made an issue for this and assigned myself: https://github.com/raiden-network/raiden/issues/5304

File: raiden/tests/scenarios/bf4_non_allowed_ops.yaml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      - parallel:
+          name: "Deposit in the other direction"
+          tasks:
+            - deposit: {from: 1, to: 0, total_deposit: 1_000_000_000_000_000_000, expected_http_status: 200}
+      - parallel:
+          name: "Assert after deposit"
+          tasks:
+            - assert: {from: 1, to: 0, total_deposit: 1_000_000_000_000_000_000, balance: 1_000_000_000_000_000_000, state: "opened"}
+      - serial:
+          name: "Try to open already opened channel, expected to not be allowed with 409"
```
Comment (LefterisJP) [inline raiden/tests/scenarios/bf4_non_allowed_ops.yaml]: This is already tested here: https://github.com/raiden-network/raiden/blob/a220f4a59604d55c6095186b0d739f3cf1e93777/raiden/tests/integration/api/test_restapi.py#L379-L384

File: raiden/tests/scenarios/bf4_non_allowed_ops.yaml [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+          tasks:
+            - open_channel: {from: 0, to: 1, total_deposit: 1_000_000_000_000_000_000, expected_http_status: 409}
+      - serial:
+          name: "Send payment of higher amount than available, expected 409"
```
Comment (LefterisJP) [inline raiden/tests/scenarios/bf4_non_allowed_ops.yaml]: This appears to not be tested.

https://codecov.io/gh/raiden-network/raiden/src/develop/raiden/api/rest.py#L1088

I will add a test for it.

File: raiden/tests/scenarios/bf4_non_allowed_ops.yaml [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+          tasks:
+            - transfer: {from: 0, to: 1, amount: 1_000_000_000_000_000_001, lock_timeout: 30, expected_http_status: 409}
+      - parallel:
+          name: "Assert that balance didn't change"
+          tasks:
+            - assert: {from: 0, to: 1, total_deposit: 1_000_000_000_000_000_000, balance: 1_000_000_000_000_000_000, state: "opened"}
+            - assert: {from: 1, to: 0, total_deposit: 1_000_000_000_000_000_000, balance: 1_000_000_000_000_000_000, state: "opened"}
+      - serial:
+          name: "Node1 tries to withdraw more than its balance"
```
Comment (LefterisJP) [inline raiden/tests/scenarios/bf4_non_allowed_ops.yaml]: This is already tested here: https://github.com/raiden-network/raiden/blob/a220f4a59604d55c6095186b0d739f3cf1e93777/raiden/tests/integration/api/test_restapi.py#L2345-L2356

File: raiden/tests/scenarios/bf4_non_allowed_ops.yaml [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+      - serial:
+          name: "Node1 withdraws it entire balance"
+          tasks:
+            - withdraw: {from: 1, to: 0, total_withdraw: 1_000_000_000_000_000_000, expected_http_status: 200}
+      - serial:
+          name: "Assert that balance changed"
+          tasks:
+            - assert: {from: 1, to: 0, total_deposit: 1_000_000_000_000_000_000, balance: 0, state: "opened"}
+      - serial:
+          name: "Node1 tries to withdraw again"
```
Comment (LefterisJP) [inline raiden/tests/scenarios/bf4_non_allowed_ops.yaml]: This and the previous are tested here:

https://github.com/raiden-network/raiden/blob/a220f4a59604d55c6095186b0d739f3cf1e93777/raiden/tests/integration/api/test_restapi.py#L2358-L2382

That test can have a bit more obvious variable names to make it obvious that the second withdraw should fail. I will change that.

File: raiden/tests/scenarios/bf4_non_allowed_ops.yaml [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+          tasks:
+            - withdraw: {from: 1, to: 0, total_withdraw: 1_000_000_000_000_000_000, expected_http_status: 409}
+      - serial:
+          name: "Assert that balance did not change"
+          tasks:
+            - assert: {from: 1, to: 0, total_deposit: 1_000_000_000_000_000_000, balance: 0, state: "opened"}
+      - serial:
+          name: "Node1 tries to deposit more than it has onchain"
```
Comment (LefterisJP) [inline raiden/tests/scenarios/bf4_non_allowed_ops.yaml]: This is tested here: https://github.com/raiden-network/raiden/blob/a220f4a59604d55c6095186b0d739f3cf1e93777/raiden/tests/integration/api/test_restapi.py#L498-L513

File: raiden/tests/scenarios/bf4_non_allowed_ops.yaml [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+          name: "Close channel"
+          tasks:
+            - close_channel: {from: 0, to: 1, expected_http_status: 200}
+      - parallel:
+          name: "Assert that channel closed"
+          tasks:
+            - assert: {from: 0, to: 1, total_deposit: 1_000_000_000_000_000_000, balance: 1_000_000_000_000_000_000, state: "closed"}
+            - assert: {from: 1, to: 0, total_deposit: 1_000_000_000_000_000_000, balance: 0, state: "closed"}
+      - serial:
+          name: "Node1 tries to deposit after channel close"
```
Comment (LefterisJP) [inline raiden/tests/scenarios/bf4_non_allowed_ops.yaml]: This is already tested here: https://github.com/raiden-network/raiden/blob/a220f4a59604d55c6095186b0d739f3cf1e93777/raiden/tests/integration/api/test_restapi.py#L897-L908

File: raiden/tests/scenarios/bf4_non_allowed_ops.yaml [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+          tasks:
+            - deposit: {from: 1, to: 0, total_deposit: 1_000_000_000_000_000_000, expected_http_status: 409}
+      - serial:
+          name: "Send payment after channel closed"
```
Comment (LefterisJP) [inline raiden/tests/scenarios/bf4_non_allowed_ops.yaml]: I don't think this is tested. I will add a test.

File: raiden/tests/scenarios/bf4_non_allowed_ops.yaml [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+          tasks:
+            - transfer: {from: 0, to: 1, amount: 1_000_000_000_000_000_001, lock_timeout: 30, expected_http_status: 409}
+      - parallel:
+          name: "Assert that balances didn't change"
+          tasks:
+            - assert: {from: 0, to: 1, total_deposit: 1_000_000_000_000_000_000, balance: 1_000_000_000_000_000_000, state: "closed"}
+            - assert: {from: 1, to: 0, total_deposit: 1_000_000_000_000_000_000, balance: 0, state: "closed"}
+      - serial: 
+          name: "Node0 tries to close the channel again"
```
Comment (LefterisJP) [inline raiden/tests/scenarios/bf4_non_allowed_ops.yaml]: Closing channel again is tested but not in the API tests. I will add a test.

File: raiden/tests/scenarios/bf4_non_allowed_ops.yaml [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+          tasks:
+            - close_channel: {from: 0, to: 1, expected_http_status: 409}
+      - parallel:
+          name: "Assert that nothing changed"
+          tasks:
+            - assert: {from: 0, to: 1, total_deposit: 1_000_000_000_000_000_000, balance: 1_000_000_000_000_000_000, state: "closed"}
+            - assert: {from: 1, to: 0, total_deposit: 1_000_000_000_000_000_000, balance: 0, state: "closed"}
+      - parallel:
+          name: "Node1 tries to open the same channel before settlement"
```
Comment (LefterisJP) [inline raiden/tests/scenarios/bf4_non_allowed_ops.yaml]: This is a funny thing to try. It's not tested. Will add a check at the same test as above.

**Comment by czepluch** [comment]
Hey @LefterisJP . Sounds good. Thanks a lot. Let just go ahead and close this PR then.

**Comment by LefterisJP** [comment]
@czepluch yep sounds good. Just let me keep it open also as a reminder until the tests mentioned are written by me.

Also thank you and @christianbrb for pointing out a list of tests for us to check and add missing ones.

**Comment by czepluch** [comment]
Alright. Just close the issue once you're done then. Thank you!

**Comment by LefterisJP** [comment]
Closing in favour or https://github.com/raiden-network/raiden/pull/5316

**Pull request closed without merging**